src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -0,0 +1,481 @@
+    previous_ = 0.0;
+
+    zero_.SetSize(true_size);
+    zero_ = 0.0;
+  }
+
+  /**
+   * @brief Set essential temperature boundary conditions (strongly enforced)
+   *
+   * @param[in] temp_bdr The boundary attributes on which to enforce a temperature
+   * @param[in] temp_bdr_coef The prescribed boundary temperature
+   */
+  template <typename TemperatureType>
+  void setTemperatureBCs(const std::set<int>& temp_bdr, TemperatureType temp)
+  {
+    // Project the coefficient onto the grid function
+    temp_bdr_coef_ = std::make_shared<mfem::FunctionCoefficient>([temp](const mfem::Vector& x, double t) -> double {
+      tensor<double, dim> x_tensor;
+      x_tensor[0] = x[0];
+      if constexpr (dim > 1) {
+        x_tensor[1] = x[1];
+      }
+      if constexpr (dim > 2) {
+        x_tensor[2] = x[2];
+      }
+
+      return temp(x_tensor, t);
+    });
+
+    bcs_.addEssential(temp_bdr, temp_bdr_coef_, temperature_);
+  }
+
+  /**
+   * @brief Advance the timestep
+   *
+   * @param[inout] dt The timestep to advance. For adaptive time integration methods, the actual timestep is returned.
+   */
+  void advanceTimestep(double& dt) override
+  {
+    temperature_.initializeTrueVec();
+
+    if (is_quasistatic_) {
+      nonlin_solver_.Mult(zero_, temperature_.trueVec());
+    } else {
+      SLIC_ASSERT_MSG(gf_initialized_[0], "Thermal state not initialized!");
+
+      // Step the time integrator
+      ode_.Step(temperature_.trueVec(), time_, dt);
+    }
+
+    temperature_.distributeSharedDofs();
+    cycle_ += 1;
+  }
+
+  template <typename MaterialType>
+  void setMaterial(MaterialType material)
+  {
+    K_functional_.AddDomainIntegral(
+        Dimension<dim>{},
+        [material]([[maybe_unused]] auto x, auto temperature) {
+          // Get the value and the gradient from the input tuple
+          auto [u, du_dx] = temperature;
+
+          auto flux = material.thermalFlux(u, du_dx);
+
+          auto source = u * 0.0;
+
+          // Return the source and the flux as a tuple
+          return serac::tuple{source, flux};
+        },
+        mesh_);
+
+    M_functional_.AddDomainIntegral(
+        Dimension<dim>{},
+        [material]([[maybe_unused]] auto x, [[maybe_unused]] auto temperature) {
+          auto [u, du_dx] = temperature;
+
+          auto source = material.cp * material.rho;
+
+          auto flux = 0.0 * du_dx;
+
+          // Return the source and the flux as a tuple
+          return serac::tuple{source, flux};
+        },
+        mesh_);
+  }
+
+  template <typename TemperatureType>
+  void setTemperature(TemperatureType temp_function)
+  {
+    // Project the coefficient onto the grid function
+    mfem::FunctionCoefficient temp_coef([temp_function](const mfem::Vector& x, double t) -> double {
+      tensor<double, dim> x_tensor;
+      x_tensor[0] = x[0];
+      if constexpr (dim > 1) {
+        x_tensor[1] = x[1];
+      }
+      if constexpr (dim > 2) {
+        x_tensor[2] = x[2];
+      }
```

Comment:
I had to do something somewhat similar in Inlet to convert to/from `mfem::Vector` - if this is something we anticipate being needed in multiple places it might be worth implementing a

```cpp
template <int dim>
tensor<double, dim> vectorToTensor(const mfem::Vector&);
```

or more broadly perhaps a tensor constructor accepting a `T*`?

Reply:
there is also 
```cpp
mfem::Vector foo = ... ;
auto x = make_tensor<dim>([&](int i){ return foo[i]; });
```

Reply:
Question: I don't think I can use `std::function` for this since it is injecting directly into the kernels. Does this need to be a type template?